[Refactor] Integrate a SettingsManager

flx-sta opened the pull request:
> [!NOTE]
> - ~~As of right now the automatic reload only works on the language change.~~
>     - ~~_Reason for that is that the whole handling of that is supposed to change anyway so there was no point in integrating it_~~
> - Also no settings-related stuff inside `game-data` was touched as I'm not fully aware of the interaction with the actual game-data object being loaded. **If anyone is better informed about that feel free to help!**

## What are the changes the user will see?

- Reload required is only shown as a `*` now with an explanation of that `*` at the bottom of the settings screen
- Opening the change-language menu doesn't reload if no change happened. It also automatically switches back to the actual language option instead of staying on `Change`
- Move touch controls now has 2 options and automatically updates it label on the fly

## Why am I making these changes?

All settings were tightly coupled with the battle-scene (as almost everything). This made it impossible to read settings before the battle-scene is ready.

## What are the changes from a developer perspective?

- All settings are now handled by the `SettingsManager` (or rather the intance of it).
- Settings are updated through a `settings.update(<category>, <key>, <value>)` method that also emits events about these changes (in `eventBus`. See `event-bus.ts` file for all the events).
- localstorage keys are now prefixed with `pky/` to make sure that no other localhost development interferes with it.
- added locales: https://github.com/Despair-Games/poketernity-locales/commit/96bf7511264b5c97348463b24a0f3d24e5a1c35f

## Screenshots/Videos

<details>
  <summary> [Videos] </summary>

https://github.com/user-attachments/assets/98b3e329-03bc-40c4-b4b1-aca4e0617c4d

https://github.com/user-attachments/assets/aab3c974-c18b-41b5-a60a-4d7454f5d966

https://github.com/user-attachments/assets/f5d561a2-c8bf-46da-9d37-1b1fcf94f565

https://github.com/user-attachments/assets/528a0b25-2df6-486d-98e6-7bf720700e79

</details>

## How to test the changes?

- Start the game
- Play around with the settings and analyze if they actually take effect!

## Checklist

- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test`)
  - ~~[ ] Have I created new automated tests (`npm run create-test`) or updated existing tests related to the PR's changes?~~
- [x] Have I provided screenshots/videos of the changes (if applicable)?

flx-sta committed:
migrate settings to settings-manager

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
move settings files into `system/settings`

flx-sta committed:
update `settings.ts`

disable more switch cases

flx-sta committed:
fix `useBoolOptions`

flx-sta committed:
minor setting adjustments

flx-sta committed:
add ui-window-type enum

flx-sta committed:
clean up settings stuff

flx-sta committed:
fix language setting related stuff

- language is updated properly
- language selection is option mode
- cursor switches back from `change` on cancel

flx-sta committed:
minor settings fixes

- rename some ls keys and store them in constnats
- set default game-speed to `2` (instead of 1)
- increase size of reload-required info text

flx-sta committed:
clean up some legacy settings leftovers

- make `uiItems` mandataory for settings ui handlers
- remove `mode` from settings ui handlers
- replace `type` with `category` in settings ui handlers
- inject settings title through category key

flx-sta committed:
clean up battle-scene

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
add confirmation for reload settings

when progress could be lost

Co-authored-by: =?UTF-8?q?Adri=C3=A1n?= <99520451+Vassiat@users.noreply.github.com>

flx-sta committed:
move `hasTouchscreen` into utils

flx-sta committed:
fix settings

- require reload now takes effect
- remove some events that couldn't be listened to anyway
- fix hide/show touchscreen only settings

flx-sta committed:
fix settings

- now reloads properly if no progress is lost
- touch controls are properly (re-) rendered
- moved color variables to css
- introduce datasets on the html element that update if uiTheme/WindowType updates
- rename `windowType` to `uiWindowType`
- touch controls can be moved again

flx-sta committed:
update submodules (locales)

flx-sta committed:
add global event bus

- remove settings manager event bus

flx-sta committed:
address comments

flx-sta commented:
https://github.com/user-attachments/assets/aab3c974-c18b-41b5-a60a-4d7454f5d966

https://github.com/user-attachments/assets/f5d561a2-c8bf-46da-9d37-1b1fcf94f565

https://github.com/user-attachments/assets/528a0b25-2df6-486d-98e6-7bf720700e79

flx-sta commented:
<img width="192" alt="image" src="https://github.com/user-attachments/assets/a21d9e50-16a5-45de-a163-49b1b18c441e" />
<img width="217" alt="image" src="https://github.com/user-attachments/assets/4df6d5e3-5e00-4200-ad6d-7a00fc4aaaa2" />

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

MokaStitcher reviewed (commented):
again stopping very early on in the review, sorry
I was trying to look at game data, the gamepad/keyboard remapping thing is such a mess wtf, so many different functions to save, called from different places '-'
I'll get back to it properly later, but for now, I think you missed to update the locale storage key here like you did in the gamepad file https://github.com/Despair-Games/poketernity/blob/4dacb5295e1265461b41ca1ec63494824a0e5b40/src/ui/settings/settings-keyboard-ui-handler.ts#L41

MokaStitcher reviewed (commented):
Made a first full pass over the code and did some limited testing.

Overall looks good, the main remarks I have so far:
- not sure if you consider the gamepad/keyboard remapping part done (I didn't look into that too closely, following the issues I found in my earlier comment)
- the game-data file needs to be cleaned up to remove everything that is no longer needed if i understand correctly? Probably settings.ts too, so that the deprecated things can be removed?
- tutorials disabled by default, is that intended?
- disabling touchpad no longer asks for confirmation


<details><summary><sub>(this is a note for me or I'll forget, don't mind me)</summary>

not introduced in this PR, but the remapping text + timer are not centered properly (screenshot is italian, but in french it goes out of the window's bounds)

![image](https://github.com/user-attachments/assets/341d1a8c-b5e3-4e1c-acf1-eccb652c65c2)

</details>

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
move settings-ui-items to `ui/settings`

flx-sta committed:
remove active language from list of languages

flx-sta committed:
make handleCancelLanguageChange return `true`

so the error sound isn't played

flx-sta committed:
remove cancel handler from language selection

flx-sta committed:
fix bypass tutorials parsing

flx-sta committed:
fix typo in Settings.ts

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
address comments

- fix gamepad being stored inside `pky/settings`
- fix gamepad settings not being loaded properly
- fix mappings now being all stored in 1 LS key
- add gamepad settings key to settings + defaults
- remove some debugging stuff

flx-sta committed:
remove obsolete todo comment

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
resolve issues after merge

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
re-impl. confirming disable touch controlls

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

MokaStitcher committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

MokaStitcher committed:
merge 'beta' into refactor/standalone-settings

MokaStitcher committed:
Merge branch 'beta' into refactor/standalone-settings

# Conflicts (related to music preference name change):
#	src/battle-scene.ts
#	src/battle.ts
#	src/data/mystery-encounters/encounters/global-trade-system-encounter.ts

MokaStitcher committed:
Merge branch 'beta' into refactor/standalone-settings

# Conflicts (experimental sprite removal):
#	src/battle-scene.ts
#	src/field/pokemon.ts
#	src/system/settings/settings.ts

MokaStitcher committed:
remove sprite set / experimental sprites settings (see PR 101)

MokaStitcher committed:
Merge branch 'beta' into refactor/standalone-settings

# Conflicts:
#	src/phases/check-switch-phase.ts
#	src/phases/damage-anim-phase.ts
#	src/phases/egg-lapse-phase.ts
#	src/phases/game-over-phase.ts
#	src/phases/level-up-phase.ts
#	src/phases/scan-ivs-phase.ts
#	src/phases/select-gender-phase.ts
#	src/phases/show-party-exp-bar-phase.ts
#	src/phases/stat-stage-change-phase.ts

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

MokaStitcher committed:
Settings PR cleanup (not for beta, see #50) (#70)

* start of cleanup

* remove redundant try/catch

* fix eslint issue

* add meta settings category for storing the settings' game version

* suggestion from code review

* make settings use the PlayerGender enum for player gender

* use player gender from settings instead of game data

* battle scene updates player sprite when the gender changes

* fix shop overlay setting

* fix error sound effect playing when changing language

* update starter select when candy upgrade setting is changed

* update money display after changing setting

* giving up on settings migration

* start cleanup of mapping settings

* reset mappings button only resets mappings for current device

* remove unused function

* fix wrap around sfx not playing in audio setting

---------

Co-authored-by: flx-sta <50131232+flx-sta@users.noreply.github.com>

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

MokaStitcher commented on src/@types/Settings.ts:
```diff
@@ -0,0 +1,124 @@
+export type SettingUiItemOption = {
+  value: number | string | boolean;
+  label: string;
+  needConfirmation?: boolean;
+  confirmationMessage?: string;
+};
```
Like we had discussed on Discord a while ago, either these lines need to go, as well as the code using them [here](https://github.com/Despair-Games/poketernity/blob/0f2d8030eef3d66ca9ce1f4184c6673ab70eaf7e/src/ui/settings/abstract-settings-ui-handler.ts#L464)
Or, this https://github.com/Despair-Games/poketernity/blob/0f2d8030eef3d66ca9ce1f4184c6673ab70eaf7e/src/ui/settings/abstract-settings-ui-handler.ts#L588 needs to go, and the [enableTouchControl ui item definition](https://github.com/Despair-Games/poketernity/blob/0f2d8030eef3d66ca9ce1f4184c6673ab70eaf7e/src/ui/settings/settings-ui-items.ts#L129) need to use those `needConfirmation` and `confirmationMessage` field (for reference here's how they were used [before](https://github.com/Despair-Games/poketernity/blob/3c14d91bb7484135dcb7daa6c3ddc0dbcf2cb5d7/src/system/settings/settings.ts#L56))

My opinion hasn't changed that I'd rather go with option 2, because the abstract settings ui handler shouldn't have code for any specific setting. Also because this way the fact that a setting needs confirmation is handled in the same place as the fact that a setting requires reloading, and it's all in one place.

But I won't fight whichever option you go with, just need to pick one or the other and clean up the rest

flx-sta replied:
I totally forgot about that discussion, sorry :D

flx-sta replied:
done

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
adopt needsConfirmation pattern

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

MokaStitcher committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

innerthunder reviewed (dismissed):
lgtm. I noticed some consts/variables with PokeRogue branding (`PRSFX` specifically), but I suppose that's something for another PR to resolve...

flx-sta commented:
> lgtm. I noticed some consts/variables with PokeRogue branding (`PRSFX` specifically), but I suppose that's something for another PR to resolve...

correct, that's something for another PR.
Afaik the art team wanted to chip in for that

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

Tempo-anon commented on src/constants.ts:
```diff
@@ -17,7 +17,22 @@ export const SAVES_ZIP_PREFIX = "poketernity_";
 
 /** Prefix for local storage keys. */
-export const LS_PREFIX = "pkty";
+export const LS_PREFIX = "pky";
```
was there a reason this was changed?

flx-sta replied:
We have a pending vote for the final decision. While that's going on I'll leave it as it is

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

MokaStitcher commented on src/system/game-speed.ts:
```diff
@@ -21,7 +19,7 @@ export function initGameSpeed() {
       return num.value;
     }
-    return thisArg.gameSpeed === 1 ? num : Math.ceil((num /= thisArg.gameSpeed));
+    return settings.settings.general.gameSpeed === 1 ? num : Math.ceil((num /= settings.settings.general.gameSpeed));
```
```suggestion
    return settings.general.gameSpeed === 1 ? num : Math.ceil((num /= settings.general.gameSpeed));
```

why did that even work

MokaStitcher commented on src/system/settings/settings-manager.ts:
```diff
@@ -0,0 +1,194 @@
+   * Getter for {@linkcode _settings}. No public setter!
+   */
+  get settings() {
+    return this._settings;
```
ah, is this why settings.settings works?
Do we need this getter since there's one for each sub category?

MokaStitcher commented on src/ui/settings/settings-keyboard-ui-handler.ts:
```diff
@@ -37,7 +38,7 @@ export default class SettingsKeyboardUiHandler extends AbstractControlSettingsUi
     this.commonSettingsCount = 0;
     this.textureOverride = "keyboard";
-    this.localStoragePropertyName = "settingsKeyboard";
+    this.localStoragePropertyName = MAPPING_CONFIG_LS_KEY;
```
```suggestion
```

ah, looks like I forgot to remove [localStoragePropertyName ](https://github.com/Despair-Games/poketernity/blob/7ff8e8d136c1a6e64841e6829ea7d1663260e62b/src/ui/settings/abstract-control-settings-ui-handler.ts#L66) from the parent class. But I'm pretty sure it's not used anymore, so might as well remove it everywhere (will need to remove the MAPPING_CONFIG_LS_KEY import too)

DayKev replied:
Uh, the "suggested" code is identical to the existing code. 🤔

MokaStitcher replied:
hu, right, thanks lol. This suggestion shouldn't be applied on its own anyway it requires touching up 3 files in total I think

MokaStitcher commented on src/ui/settings/settings-ui-items.ts:
```diff
@@ -0,0 +1,374 @@
+      {},
+      { requiresConfirmation: true, confirmationMessage: t("settings:confirmDisableTouch") },
+    ),
+    //TODO: needs confirmation for `disabled`
```
```suggestion
```

should be good right?

MokaStitcher committed:
Merge branch 'beta' into refactor/standalone-settings

# Conflicts:
#	src/battle-scene.ts
#	src/data/mystery-encounters/encounters/global-trade-system-encounter.ts
#	src/field/anims.ts
#	src/phases/summon-phase.ts

MokaStitcher reviewed (commented):
A few notes. Getting there.
Time to playtest a bunch :D

MokaStitcher commented on src/system/settings/default-settings.ts:
```diff
@@ -0,0 +1,80 @@
+  expGainsSpeed: ExpGainsSpeed.DEFAULT,
+  partyExpNotificationMode: ExpNotification.DEFAULT,
+  skipSeenDialogues: false,
+  eggSkipPreference: EggSkipPreference.NEVER,
```
```suggestion
  eggSkipPreference: EggSkipPreference.ASK,
```
Asking was the default [before](https://github.com/Despair-Games/poketernity/blob/0c56fc32b467de4aa12ba6ef982d1b351daa0cb7/src/system/settings/settings.ts#L320)

MokaStitcher reviewed (commented):
Testing went well, once the issues in this review and the previous one are fixed i'll be able to approve fast!

edit (x2): nevermind ~~one more~~ couple more issues
- These are used as the locale storage key for custom touch control positions: https://github.com/Despair-Games/poketernity/blob/0b6d618d05d9e10f1d850d1f8ddf2f427d0cb122/src/ui/settings/move-touch-controls-handler.ts#L6-L7 They should probably also use the pky prefix, and if possible not be saved in their own individual entry like they are now? Maybe that second part can be for another time but since this is settings related and other settings like mappings had their keys changed, it would make sense to have it done in this PR

- Another local storage key that didn't get cleaned up here:
https://github.com/Despair-Games/poketernity/blob/0b6d618d05d9e10f1d850d1f8ddf2f427d0cb122/src/system/game-data.ts#L73-L74
Although the settings one is not used, we might as well return the correct key, or add a TODO to update this whole thing


Final thing, we also should merge that locales PR https://github.com/Despair-Games/poketernity-locales/pull/13

flx-sta committed:
Update src/system/settings/default-settings.ts

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

flx-sta committed:
Clean up move-touch-controls-handlers

- Optimize element selections
- Add missing translations
- Make use of getters
- Make use of phasers integrated orientation detection
- Update ls keys to match new pattern

flx-sta committed:
Update LS key in `game-data.ts` `getDataTypeKey`

flx-sta committed:
Apply suggestions

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
remove settings.settings

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
Fix `settingsManager[this.category]` calls

MokaStitcher commented on src/system/game-data.ts:
```diff
@@ -700,106 +671,44 @@ export class GameData {
     let tutorials: object = {};
-    if (localStorage.hasOwnProperty(key)) {
-      tutorials = JSON.parse(localStorage.getItem(key)!); // TODO: is this bang correct?
+    const lsItem = localStorage.getItem(key);
```
uh, had forgotten about this change, it's pretty unrelated to the PR and will cause conflicts with #270, but no matter, unless you feel like reverting it

flx-sta replied:
I think it's fine. Whoever comes first will easily resolve it

MokaStitcher commented on src/ui/settings/move-touch-controls-handler.ts:
```diff
@@ -45,41 +42,139 @@ export default class MoveTouchControlsHandler {
-        this.changeOrientation(screenSize.width > screenSize.height);
-      }
+    this.initListeners();
+    console.log("MoveTouchControlsHandler currentOrientation:", this.currentOrientation);
```
```suggestion
```

MokaStitcher commented on src/constants.ts:
```diff
@@ -17,10 +17,25 @@ export const SAVES_ZIP_PREFIX = "poketernity_";
+/** Key for the local storage item storing the settings. */
+export const SETTINGS_LS_KEY = `${LS_PREFIX}/settings`;
+
+/** Key prefix for the local storage item storing the users data. */
+export const DATA_LS_KEY_PREFIX = `${LS_PREFIX}/data`;
+
```
```suggestion
```

What is this for? It doesn't seem to be used, better leave it for a later PR that will handle updating all local storage keys

MokaStitcher commented on src/constants.ts:
```diff
@@ -17,10 +17,25 @@ export const SAVES_ZIP_PREFIX = "poketernity_";
 
 /** Prefix for local storage keys. */
-export const LS_PREFIX = "pkty";
+export const LS_PREFIX = "pky";
```
```suggestion
export const LS_PREFIX = "pkty";
```
adding it since i made other suggestions so you don't need a separate commit

MokaStitcher commented on src/ui/settings/move-touch-controls-handler.ts:
```diff
@@ -109,52 +204,37 @@ export default class MoveTouchControlsHandler {
+    this.resetBtn.addEventListener("click", () => {
       this.resetPositions();
+      eventBus.emit("touchControls/move/reset");
+      eventBus.emit("touchControls/move/end");
```
```suggestion
```

I don't think the reset option should emit the end event, since it doesn't close the view

flx-sta replied:
good point

MokaStitcher commented on src/ui/settings/settings-ui-items.ts:
```diff
@@ -0,0 +1,373 @@
+    touchscreenOnly: true,
+  },
+  {
+    key: "enableVibration",
```
if I could have one final nit, it would be to put the enableVibration setting above the enable touch control one, as it was before

MokaStitcher reviewed (commented):
few final nits. Looks good!
I merged the remaining 2 locale PRs so you can go ahead and update the submodule in this PR (if you use command line for it: `git submodule update --remote --recursive --force` )

flx-sta committed:
Update src/ui/settings/move-touch-controls-handler.ts

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
Reset `LS_PREFIX` to `pkty`

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
remove debug log

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
remove `DATA_LS_KEY_PREFIX` for now

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
move `"enableVibration"` above `"moveTouchControls"`

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

flx-sta committed:
move `enableVibration` above `enableTouchControls`

now its
- enableVibration
- enableTouchControls
- moveTouchControls

flx-sta committed:
sync locales repo

MokaStitcher reviewed (dismissed):
🚀🚀🚀

PigeonBar commented on src/ui-inputs.ts:
```diff
@@ -216,23 +216,13 @@ export class UiInputs {
-        Setting[settingGameSpeed].options.findIndex((item) => item.label === `${globalScene.gameSpeed}x`) + 1,
-      );
+    if (up && settings.general.gameSpeed < 5) {
+      settings.update("general", "gameSpeed", settings.general.gameSpeed + 1);
```
This approach causes edge cases when pressing the "speed up" or "speed down" keys while at a fractional game speed.

https://github.com/user-attachments/assets/23f82016-008a-49b5-a8a1-0b17bee23774

(Most likely, we'll have to replace these with separate functions to search up the previous or next speed option.)

flx-sta replied:
Good catch

flx-sta replied:
Fixed: 3a55ba0

flx-sta committed:
fix game-speed quick-buttons update

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings

MokaStitcher commented on src/ui-inputs.ts:
```diff
@@ -216,23 +217,17 @@ export class UiInputs {
   }
 
   buttonSpeedChange(up = true): void {
```
Proposed rewrite of the function (untested). I don't like that this is calling the uiHandler's `show` function and would rather just **NOT** allow changing speed with the shortcuts while in the settings. Also this would only update the setting if there is a change

```ts
      buttonSpeedChange(up = true): void {
        if (globalScene.ui?.getMode() === Mode.SETTINGS) {
          return;
        }
        const index = gameSpeeds.findIndex((n) => n === settings.general.gameSpeed) ?? 0;
        let newIndex = up ? Math.min(index + 1, gameSpeeds.length - 1) : Math.max(index - 1, 0);
        if (newIndex !== index) {
          settings.update("general", "gameSpeed", gameSpeeds[newIndex]);
        }
      }
```

flx-sta replied:
~~@MokaStitcher doesn’t this eliminate the whole purpose of the “quick access” of those buttons ? 
I’d rather keep them as they are. But I’m open for reasons to apply your change~~

mb, I missed the not

MokaStitcher replied:
The quick buttons are meant to be used during gameplay, not while in the settings, that makes absolutely no sense to me. Why would you want to change game speed with a shortcut no one knows about while being in the audio settings menu for example?
<sub>I'd also argue that no one uses them either way.

I don't know that show is basically re-rendering the whole settings I don't see a world where it wouldn't cause issues to do that while you're navigating said settings. But I guess I should just try it and see if that's the case

MokaStitcher replied:
if you don't want to change that, fine but can you at least move the duplicate 
```
settings.update("general", "gameSpeed", gameSpeeds[nextIndex]);
      if (globalScene.ui?.getMode() === Mode.SETTINGS) {
        (globalScene.ui.getHandler() as SettingsUiHandler).show([]);
      }
```
outside of the if/else, and only execute it if nextIndex is not the same as the current one? (also I just noticed it was `if (up) ... else if (!up)` instead of just `if (up) ... else ...` lol)

MokaStitcher replied:
Here's an argument against allowing the shortcuts' use in the settings:
if you press the shortcut button because you want to map it to another action, it also calls the function and updates the speed
![image](https://github.com/user-attachments/assets/c6c9d7e0-238d-489a-9f37-7b5974964d07)

I didn't find other issues with calling show itself, aside from the fact that if you're in the General tab it scrolls you back to the top of the tab, which is a bit weird but fine

flx-sta replied:
mb, I missed the `not`

flx-sta committed:
prevent quick gamespeed change in settings

- move and rename `GAME_SPEEDS` constant

flx-sta committed:
fix game-speeds import in settings-ui-items

flx-sta committed:
Merge branch 'beta' into refactor/standalone-settings
